File: backend/models/WalletMember.js
const mongoose = require('mongoose');

const WalletMemberSchema = new mongoose.Schema({
  walletId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Wallet',
    required: true
  },
  userId: { 
    type: String, 
    required: true
  },
  permission: { 
    type: String, 
    required: true,
    enum: ['read', 'write', 'owner'],
    default: 'read'
  },
  joinedAt: { 
    type: Date, 
    default: Date.now 
  }
}, {
  timestamps: true
});

// Índice composto para garantir que um usuário só pode ter uma relação por wallet
WalletMemberSchema.index({ walletId: 1, userId: 1 }, { unique: true });
WalletMemberSchema.index({ userId: 1 });
WalletMemberSchema.index({ walletId: 1 });

const getWalletMemberModel = () => {
  const collectionName = 'wallet_members';
  const modelName = 'WalletMember'; // Nome usado no ref (se necessário)
  
  // Verificar se já existe com o nome do modelo
  if (mongoose.models[modelName]) {
    return mongoose.models[modelName];
  }
  
  // Verificar se existe com o nome da collection
  if (mongoose.models[collectionName]) {
    return mongoose.models[collectionName];
  }
  
  // Registrar com o nome do modelo
  return mongoose.model(modelName, WalletMemberSchema, collectionName);
};

module.exports = { WalletMemberSchema, getWalletMemberModel };

File: lib/models/wallet_member.dart
class WalletMember {
  final String walletId;
  final String userId;
  final String permission; // 'read', 'write', 'owner'
  final DateTime joinedAt;
  final String? email;
  final String? name;
  final bool isOwner;

  WalletMember({
    required this.walletId,
    required this.userId,
    required this.permission,
    required this.joinedAt,
    this.email,
    this.name,
    this.isOwner = false,
  });

  Map<String, dynamic> toJson() {
    return {
      'walletId': walletId,
      'userId': userId,
      'permission': permission,
      'joinedAt': joinedAt.toIso8601String(),
      'email': email,
      'name': name,
      'isOwner': isOwner,
    };
  }

  factory WalletMember.fromJson(Map<String, dynamic> json) {
    DateTime parseDate(dynamic dateValue) {
      if (dateValue is String) {
        return DateTime.parse(dateValue);
      } else if (dateValue is Map) {
        final dateStr = dateValue['\$date'] as String? ?? dateValue.toString();
        return DateTime.parse(dateStr);
      } else {
        return dateValue as DateTime;
      }
    }

    return WalletMember(
      walletId: json['walletId']?.toString() ?? '',
      userId: json['userId'] as String? ?? '',
      permission: json['permission'] as String? ?? 'read',
      joinedAt: parseDate(json['joinedAt']),
      email: json['email'] as String?,
      name: json['name'] as String?,
      isOwner: json['isOwner'] as bool? ?? false,
    );
  }
}

